### ISSUE TITLE
Unnecessary computation of hash for BlockHeader and ShardChunkHeader

### ISSUE THREAD

**Issue by bowenwang1996** [near/nearcore#4581]
When we deserialize `BlockHeader` or `ShardChunkHeader`, we always end up recomputing the hash through `#[borsh(init)]`, which is fine for data received from network but is completely wasteful for reading from storage and likely contributes to some rpc slowdowns that we've observed.

**Comment by marcelo-gonzalez** [comment]
So it seems there are actually many different spots where these hashes are recomputed, with various difficulties in untangling. The kind that would probably be easiest to undo now are the ones of the form:


```rust
#[borsh_init(compute_and_set_the_hash)]
struct Foo {
       data: Data,

       #[borsh_skip]
       hash: CryptoHash,
}

fn get_foo(hash: &CryptoHash) -> Result<&Foo> {
   ...
}
```

In this case, we have the hash in hand, so we can just copy it instead of recomputing it. But then there's the trickier case:

```rust
struct Foo {
       data: Data,

       has_its_own_borsh_skipped_hash: Bar,

       #[borsh_skip]
       hash: CryptoHash,
}
```

In this case, we have the `struct Foo` hash, but not the inner hashes, so we kind of have no choice but to recompute those. Couple examples of this are the `ShardChunkHeader`s in `Block` and the `SignedTransaction`s in `ShardChunk`. Only way I can think of to untangle those cases is to migrate the database so that the hashes are included in the values on-disk. But maybe it's best to just try to tackle the first case for now and see how much that helps things, since I would hesitate to add new migrations that might be a pain and take quite a long time to complete without at least taking the low hanging fruit first.

As far as actually implementing the first, easier case, the simple thing to do could be to just remove the #[borsh_init] and manually either compute or copy the hash according to the situation. But that would require hunting down every case and making sure we haven't forgotten about anything. Also anyone adding new types containing a BlockHeader would have to somehow realize that the hash needs to get taken care of manually. Maybe a big bold comment could do that but I think it would be nice to still have the compiler help guarantee that the hashes are set. Could be done w something like this?

```rust
#[derive(BorshSerialize, BorshDeserialize)]
pub struct BlockHeaderDataV1 {
    pub prev_hash: CryptoHash,
    pub inner_lite: BlockHeaderInnerLite,
    pub inner_rest: BlockHeaderInnerRest,
    pub signature: Signature,
    // no hash
}

pub struct BlockHeaderDataV2 {...}
pub struct BlockHeaderDataV3 {...}

pub enum BlockHeaderData {
    BlockHeaderDataV1(...)
    ...
}

// no borsh impls
pub struct BlockHeaderV1 {
    data: BlockHeaderDataV1
    hash: CryptoHash
}

impl BlockHeader {
     fn from_data(d: BlockHeaderData) {
     	// compute hash
     }

     fn from_hashed_data(d: BlockHeaderData, h: &CryptoHash) {
     	// copy hash
     }
}
```

(Also maybe instead of the `data: BlockHeaderDataV1` field, could just re-list all the fields for clarity. Just a matter of style I guess) Here, in every place where we serialize or deserialze a block header we would replace the type with BlockHeaderData, relying on the borsh serialization being the same when removing the `borsh_skip`ed fields. Downside is that we'd have to do the same splitting of every struct containing a `BlockHeader` that's serialized or deserialized anywhere, e.g.:

```rust
pub struct BlockDataV1 {
    pub header: BlockHeaderData
    // all other fields the same as in struct BlockV1
}

pub enum BlockData {
    BlockHeaderDataV1(...)
    ...
}
   
impl Block {
     fn from_data(d: BlockData) {
     	// compute hash
     }

     fn from_hashed_data(d: BlockData, h: &CryptoHash) {
     	// copy hash
     }
}
```

But at least it would be difficult to mess up since we'd remove the `#[derive(BorshSerialize, BorshDeserialize)]` from `enum BlockHeader`, so that any place where we forget about this in a struct that cares about borsh, we should get a compile error.

@bowenwang1996 any thoughts on this approach?

**Comment by bowenwang1996** [comment]
@marcelo-gonzalez I think we can go with the approach where we introduce a new struct under the enum and remove `#[derive(BorshSerialize, BorshDeserialize)]` and make sure for the new struct, the hash is not recomputed when loading from storage. For historical data, we can leave it there for now and decide later whether we want to do a database migration.

**Comment by marcelo-gonzalez** [comment]
> I think we can go with the approach where we introduce a new struct under the enum

@bowenwang1996 here do you mean something like the above duplication into `BlockHeaderData` and `BlockHeader` that are the same except for the hash + borsh derives or something else? because I see a couple different things you could have meant by new struct under the enum

**Comment by bowenwang1996** [comment]
> here do you mean something like the above duplication into BlockHeaderData and BlockHeader that are the same except for the hash + borsh derives or something else?

Yes. What other ideas do you have?

**Comment by marcelo-gonzalez** [comment]
@bowenwang1996 ok sounds good. The others are kind of just variants on that, but I think less clear. So I'll just go with that

**Comment by marcelo-gonzalez** [comment]
So after implementing something like the above, I tried to take a look at how much work is actually saved by this. At least on my computer, hashing block headers in a release build usually takes around 2 microseconds and hashing chunk headers takes somewhere around 500 nanos. Given that this hash computation only happens when the object is not cached, and we're reading it from disk, those 2 microseconds are in the context of latency already in the 10s to 100s of micros. Especially in the case where the read comes as part of several (for example, getting a block at a particular height involves reading from the `ColBlockHeight` column first, then reading the block), the time we'd save by not recomputing the hash is way dominated by the IO

Given those numbers, I'm not really convinced that either of the two approaches I outlined above is worth it... In the case of the approach I thought was best (the one where we define a different `enum BlockData`, deserialize into that, then move it into the final `enum Block`), the downsides are lots of code churn, some duplication in the new type definitions, and an eventual `memcpy` into the new structure that in the case of Chunks, which take so little time to hash anyway, already almost wipes out the small performance gain. In the case of the other approach (just removing `#[borsh_init]` and manually either recomputing the hash or copying it depending on the situation), the downside is the correctness concern (e.g. what if we forget to set the hash somewhere...).

@bowenwang1996 is this issue somethign we've seen directly cause trouble anywhere, or maybe my analysis here is not considering something?
